File: dashboard/ios/AttobotDashboard/Core/Format.swift
import Foundation

/// Verbatim port of dashboard/web/src/format.ts / dashboard/mobile/src/lib/format.ts.
/// Pure helpers shared across screens; `Date()` and `DateFormatter` replace the
/// JS `Date`/`toLocaleString` used in the RN client.

enum Format {
    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    /// Parse an ISO8601 string (with or without fractional seconds / timezone).
    static func parseISO(_ iso: String) -> Date? {
        if let d = isoFormatter.date(from: iso) { return d }
        let fallback = ISO8601DateFormatter()
        if let d = fallback.date(from: iso) { return d }
        // Fallback for "Z-less" RFC3339 with offset, e.g. 2024-01-01T00:00:00+02:00
        let alt = ISO8601DateFormatter()
        alt.formatOptions = [.withInternetDateTime]
        return alt.date(from: iso)
    }

    static func timeAgo(_ iso: String?) -> String {
        guard let iso, !iso.isEmpty else { return "—" }
        guard let t = parseISO(iso) else { return iso }
        let s = max(0, floor(Date().timeIntervalSince(t)))
        if s < 5 { return "just now" }
        if s < 60 { return "\(Int(s))s ago" }
        let m = Int(s) / 60
        if m < 60 { return "\(m)m ago" }
        let h = m / 60
        if h < 24 { return "\(h)h ago" }
        let d = h / 24
        return "\(d)d ago"
    }

    static func formatDateTime(_ iso: String?) -> String {
        guard let iso, !iso.isEmpty else { return "—" }
        guard let t = parseISO(iso) else { return iso }
        return Self.dateTimeFormatter.string(from: t)
    }

    private static let dateTimeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateStyle = .medium
        f.timeStyle = .medium
        return f
    }()

    /// Format a byte count. nil → "—"; <1024 "N B"; then KB/MB/GB/TB, 1 decimal
    /// unless the value is >=100 (then 0 decimals). Mirrors RN `formatBytes`.
    static func formatBytes(_ n: Int64?) -> String {
        guard let n else { return "—" }
        return formatBytesNum(Double(n))
    }

    private static func formatBytesNum(_ num: Double) -> String {
        guard !num.isNaN else { return "—" }
        if num < 1024 { return "\(Int(num)) B" }
        let units = ["KB", "MB", "GB", "TB"]
        var v = num / 1024
        var i = 0
        while v >= 1024 && i < units.count - 1 {
            v /= 1024
            i += 1
        }
        let decimals = v >= 100 ? 0 : 1
        return String(format: "%.\(decimals)f %@", v, units[i])
    }

    /// Format a millisecond duration.
    static func formatMs(_ ms: Double?) -> String {
        guard let ms, !ms.isNaN else { return "—" }
        if ms < 1000 { return "\(Int(ms)) ms" }
        return String(format: "%.1f s", ms / 1000)
    }

    /// Collapse whitespace, trim; if longer than n, first n chars + ellipsis.
    static func truncate(_ s: String?, n: Int = 100) -> String {
        guard let s, !s.isEmpty else { return "" }
        let collapsed = s.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if collapsed.count > n {
            let end = collapsed.index(collapsed.startIndex, offsetBy: n)
            return String(collapsed[..<end]) + "…"
        }
        return collapsed
    }
}

File: dashboard/mobile/src/lib/format.ts
// Verbatim copy of dashboard/web/src/format.ts — pure formatting helpers shared
// across screens. (toLocaleString / Date.now() both work under React Native.)

export function timeAgo(iso: string | null | undefined): string {
  if (!iso) return "—";
  const t = new Date(iso).getTime();
  if (Number.isNaN(t)) return iso;
  const s = Math.max(0, Math.floor((Date.now() - t) / 1000));
  if (s < 5) return "just now";
  if (s < 60) return `${s}s ago`;
  const m = Math.floor(s / 60);
  if (m < 60) return `${m}m ago`;
  const h = Math.floor(m / 60);
  if (h < 24) return `${h}h ago`;
  const d = Math.floor(h / 24);
  return `${d}d ago`;
}

export function formatDateTime(iso: string | null | undefined): string {
  if (!iso) return "—";
  const t = new Date(iso).getTime();
  if (Number.isNaN(t)) return iso;
  return new Date(iso).toLocaleString();
}

export function formatBytes(n: number | string | null | undefined): string {
  const num = typeof n === "string" ? parseInt(n, 10) : n;
  if (num == null || Number.isNaN(num)) return "—";
  if (num < 1024) return `${num} B`;
  const units = ["KB", "MB", "GB", "TB"];
  let v = num / 1024;
  let i = 0;
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024;
    i++;
  }
  return `${v.toFixed(v >= 100 ? 0 : 1)} ${units[i]}`;
}

export function formatMs(ms: number | null | undefined): string {
  if (ms == null) return "—";
  if (ms < 1000) return `${ms} ms`;
  return `${(ms / 1000).toFixed(1)} s`;
}

// Truncate a SQL query / long string for compact cells; full text kept available.
export function truncate(s: string | null | undefined, n = 100): string {
  if (!s) return "";
  const one = s.replace(/\s+/g, " ").trim();
  return one.length > n ? one.slice(0, n) + "…" : one;
}
